src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -1953,6 +1953,67 @@ void M2(object o)
             );
         }
 
+        [Fact, WorkItem(32879, "https://github.com/dotnet/roslyn/issues/32879")]
+        public void ThrowExpressionInNullCoalescingOperator()
+        {
+            var source = @"
+class C
+{
+    string Field = string.Empty;
+    void M(C? c)
+    {
+        _ = c ?? throw new System.ArgumentNullException(nameof(c));
+        c.ToString();
+    }
+    void M2(C? c)
+    {
+        _ = c?.Field ?? throw new System.ArgumentNullException(nameof(c));
+        c.ToString();
+        c.Field.ToString();
+    }
+    void M3(C? c)
+    {
+        _ = c ?? throw new System.ArgumentNullException(c.ToString()); // 1
+    }
+}
+";
+            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
+                // (18,57): warning CS8602: Possible dereference of a null reference.
+                //         _ = c ?? throw new System.ArgumentNullException(c.ToString()); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(18, 57)
+                );
+        }
+
+        [Fact, WorkItem(32877, "https://github.com/dotnet/roslyn/issues/32877")]
+        public void CheckReceiverOfThrow()
+        {
+            var source = @"
+class C
+{
+    void M(System.Exception? e, C? c)
+    {
+        _ = c ?? throw e;
+        throw null; // ok
```

Comment:
>throw null; // ok [](start = 8, length = 17)

What is the rational behind handling difference in handling of ```throw null``` and ```throw e```?

Reply:
We could warn about `throw null` too, but then we'll want to allow suppressing it. Our own tests become littered with `throw null!`. 

Given this, my rationalization for not warning on `throw null` is that it is extremely close to explicitly throwing NRE. 
I understand that the user didn't literally write `throw new NRE();`, but I'm treating it as such.

In short, I think we're not helping users by warning on `throw null` (low or even negative value).

Reply:
Is there anywhere else where we take that approach? Using the "our own tests become littered with `throw null!` is bad reasoning IMO, and we should be deliberate and consistent in our implementation of NRT.

---
In reply to: [252794573](https://github.com/dotnet/roslyn/pull/32991#discussion_r252794573) [](ancestors = 252794573)

Reply:
@333fred Are we helping someone catch a problem in their code (ie. an unexpected NRE) by warning on `throw null;` or are we annoying them?
My observation about our codebase is that firsthand experience leads me to think the latter. That said, I recognize that it is not clear cut.

Reply:
I do get where you're coming from with that argument, but given that this _does_ cause an NRE, and the fix to show intention on the programmer's part is extremely simple (just adding a `!`), I think it would be better to warn. The number of people who do this will be small, and I think it would be better to have them show that this was intentional by adding the suppression.

---
In reply to: [253555590](https://github.com/dotnet/roslyn/pull/32991#discussion_r253555590) [](ancestors = 253555590)